File: frontend/lib/platformSearch.ts
/**
 * JobSearch.ai — Platform Search URL Builder
 * ============================================
 * Generates real job-platform search URLs from user criteria.
 * 100% client-side. No backend call. No API key. No scraping.
 *
 * These URLs open the platform's OWN search results page in a new tab.
 *
 * INVARIANTS:
 * - Never uses Google/Bing search as a fallback.
 * - Never invents job IDs or individual job posting URLs.
 * - Never stores these as individual job_url values.
 * - Clicking Browse NEVER changes application status.
 * - If a platform's real URL format is unknown, returns null (button disabled).
 */

export interface SearchCriteria {
  role: string;
  location: string;
  experience?: string;
  skills?: string[];
  remote?: boolean;
  internship?: boolean;
  fullTime?: boolean;
}

export interface PlatformSearchConfig {
  platform: string;
  displayName: string;
  icon: string;
  /**
   * Build the real search URL from user criteria.
   * Returns null if criteria cannot produce a valid known URL for this platform.
   */
  buildUrl: (criteria: SearchCriteria) => string | null;
}

/** URL-encodes a string safely. */
const enc = (s: string) => encodeURIComponent(s.trim());

/**
 * Platform search URL configurations.
 * Each entry uses the platform's documented public search URL format.
 */
export const PLATFORM_CONFIGS: PlatformSearchConfig[] = [
  {
    platform: "linkedin",
    displayName: "LinkedIn",
    icon: "💼",
    buildUrl: ({ role, location, remote }) => {
      if (!role) return null;
      const params = new URLSearchParams({ keywords: role, location: location || "" });
      if (remote) params.set("f_WT", "2"); // 2 = Remote filter
      return `https://www.linkedin.com/jobs/search/?${params.toString()}`;
    },
  },
  {
    platform: "indeed",
    displayName: "Indeed",
    icon: "🔎",
    buildUrl: ({ role, location, remote }) => {
      if (!role) return null;
      const params = new URLSearchParams({ q: role, l: location || "" });
      if (remote) params.set("remotejob", "032b3046-06a3-4876-8dfd-474eb5e7ed11");
      // Use in.indeed.com for India, otherwise global
      const base =
        (location || "").toLowerCase().includes("india") || !location
          ? "https://in.indeed.com/jobs"
          : "https://www.indeed.com/jobs";
      return `${base}?${params.toString()}`;
    },
  },
  {
    platform: "naukri",
    displayName: "Naukri",
    icon: "🧑‍💻",
    buildUrl: ({ role, location }) => {
      if (!role) return null;
      // Naukri URL: /role-slug-jobs?k=role&l=location
      const roleSlug = role.toLowerCase().replace(/\s+/g, "-").replace(/[^a-z0-9-]/g, "");
      const params = new URLSearchParams({ k: role, l: location || "" });
      return `https://www.naukri.com/${roleSlug}-jobs?${params.toString()}`;
    },
  },
  {
    platform: "internshala",
    displayName: "Internshala",
    icon: "🎓",
    buildUrl: ({ role, internship }) => {
      if (!role) return null;
      const roleSlug = role.toLowerCase().replace(/\s+/g, "-").replace(/[^a-z0-9-]/g, "");
      // Internshala uses /internships or /jobs depending on type
      if (internship) {
        return `https://internshala.com/internships/keywords-${enc(roleSlug)}/`;
      }
      return `https://internshala.com/jobs/keywords-${enc(roleSlug)}/`;
    },
  },
  {
    platform: "unstop",
    displayName: "Unstop",
    icon: "🏆",
    buildUrl: ({ role, location }) => {
      if (!role) return null;
      const params = new URLSearchParams({ search: role });
      if (location) params.set("location", location);
      return `https://unstop.com/jobs?${params.toString()}`;
    },
  },
  {
    platform: "foundit",
    displayName: "Foundit",
    icon: "🔍",
    buildUrl: ({ role, location }) => {
      if (!role) return null;
      const params = new URLSearchParams({
        searchType: "personalizedSearch",
        keyword: role,
        location: location || "",
      });
      return `https://www.foundit.in/search/jobs?${params.toString()}`;
    },
  },
  {
    platform: "monster",
    displayName: "Monster",
    icon: "👾",
    buildUrl: ({ role, location }) => {
      if (!role) return null;
      const params = new URLSearchParams({ query: role, locations: location || "" });
      return `https://www.monsterindia.com/srp/results?${params.toString()}`;
    },
  },
  {
    platform: "wellfound",
    displayName: "Wellfound",
    icon: "🚀",
    buildUrl: ({ role, location }) => {
      if (!role) return null;
      const params = new URLSearchParams({ role, location: location || "" });
      return `https://wellfound.com/jobs?${params.toString()}`;
    },
  },
  {
    platform: "glassdoor",
    displayName: "Glassdoor",
    icon: "🪟",
    buildUrl: ({ role, location }) => {
      if (!role) return null;
      // Glassdoor India
      const params = new URLSearchParams({ "sc.keyword": role });
      if (location) params.set("locT", "N");
      return `https://www.glassdoor.co.in/Job/jobs.htm?${params.toString()}`;
    },
  },
  {
    platform: "dice",
    displayName: "Dice",
    icon: "🎲",
    buildUrl: ({ role, location, remote }) => {
      if (!role) return null;
      const params = new URLSearchParams({ q: role });
      if (location) params.set("location", location);
      if (remote) params.set("filters.workplaceTypes", "Remote");
      return `https://www.dice.com/jobs?${params.toString()}`;
    },
  },
];

export interface PlatformLink {
  platform: string;
  displayName: string;
  icon: string;
  searchUrl: string | null;   // null = disabled (unknown URL format)
  available: boolean;
  label: string;              // "Browse LinkedIn Jobs"
}

/**
 * Generates platform search shortcut links from user criteria.
 * Pure client-side — no backend call, no API, no scraping.
 *
 * @returns Array of PlatformLink, available=false when URL cannot be determined.
 */
export function buildPlatformLinks(criteria: SearchCriteria): PlatformLink[] {
  return PLATFORM_CONFIGS.map((config) => {
    const url = config.buildUrl(criteria);
    return {
      platform: config.platform,
      displayName: config.displayName,
      icon: config.icon,
      searchUrl: url,
      available: url !== null,
      label: `Browse ${config.displayName} Jobs`,
    };
  });
}

File: backend/app/job_discovery/platform_search.py
"""
Platform Search Shortcut Generator
====================================
Generates real platform job-search URLs based on user criteria.

These are NOT individual job listings.
These are direct links to the platform's own search results page,
pre-populated with the user's query (role, location, experience, etc.).

Rules:
- Never fabricate individual job URLs.
- Never claim JobSearch.ai retrieved full platform inventory.
- Never call any platform API.
- Clicking Browse opens the platform's REAL search results in a new tab.
- Opening the link does NOT change any application status.
"""

from urllib.parse import urlencode


PLATFORM_CONFIGS = [
    {
        "platform": "linkedin",
        "display_name": "LinkedIn",
        "icon": "💼",
        "base_url": "https://www.linkedin.com/jobs/search/",
        "params": lambda role, location, remote, experience: {
            "keywords": role,
            "location": location,
            **({"f_WT": "2"} if remote else {}),  # 2 = Remote
        },
    },
    {
        "platform": "indeed",
        "display_name": "Indeed",
        "icon": "🔎",
        "base_url": "https://in.indeed.com/jobs",
        "params": lambda role, location, remote, experience: {
            "q": role,
            "l": location,
            **({"remotejob": "032b3046-06a3-4876-8dfd-474eb5e7ed11"} if remote else {}),
        },
    },
    {
        "platform": "naukri",
        "display_name": "Naukri",
        "icon": "🧑‍💻",
        "base_url": "https://www.naukri.com/jobs-in-india",
        "params": lambda role, location, remote, experience: {
            "k": role,
            "l": location,
        },
    },
    {
        "platform": "internshala",
        "display_name": "Internshala",
        "icon": "🎓",
        "base_url": "https://internshala.com/jobs/",
        "params": lambda role, location, remote, experience: {
            "keywords": role,
        },
    },
    {
        "platform": "unstop",
        "display_name": "Unstop",
        "icon": "🏆",
        "base_url": "https://unstop.com/jobs",
        "params": lambda role, location, remote, experience: {
            "search": role,
            "location": location,
        },
    },
    {
        "platform": "foundit",
        "display_name": "Foundit",
        "icon": "🔍",
        "base_url": "https://www.foundit.in/search/jobs",
        "params": lambda role, location, remote, experience: {
            "searchType": "personalizedSearch",
            "keyword": role,
            "location": location,
        },
    },
    {
        "platform": "monster",
        "display_name": "Monster",
        "icon": "👾",
        "base_url": "https://www.monsterindia.com/srp/results",
        "params": lambda role, location, remote, experience: {
            "query": role,
            "locations": location,
        },
    },
    {
        "platform": "wellfound",
        "display_name": "Wellfound",
        "icon": "🚀",
        "base_url": "https://wellfound.com/jobs",
        "params": lambda role, location, remote, experience: {
            "role": role,
            "location": location,
        },
    },
    {
        "platform": "glassdoor",
        "display_name": "Glassdoor",
        "icon": "🪟",
        "base_url": "https://www.glassdoor.co.in/Job/jobs.htm",
        "params": lambda role, location, remote, experience: {
            "sc.keyword": role,
        },
    },
    {
        "platform": "dice",
        "display_name": "Dice",
        "icon": "🎲",
        "base_url": "https://www.dice.com/jobs",
        "params": lambda role, location, remote, experience: {
            "q": role,
            "location": location,
        },
    },
]


def build_platform_search_url(config: dict, role: str, location: str, remote: bool, experience: str) -> str:
    """Constructs a real platform search URL from user criteria."""
    params = config["params"](role, location, remote, experience)
    # Remove empty values
    params = {k: v for k, v in params.items() if v}
    return f"{config['base_url']}?{urlencode(params)}"


def generate_platform_links(
    role: str,
    location: str = "India",
    remote: bool = False,
    experience: str = "",
    internship: bool = False,
) -> list[dict]:
    """
    Returns a list of platform search shortcuts for the given criteria.
    Each entry is a dict with:
      - platform, display_name, icon, search_url, label
    None of these are individual job listings.
    """
    links = []
    for config in PLATFORM_CONFIGS:
        # Internshala is primarily for internships — always include it
        url = build_platform_search_url(config, role, location, remote, experience)
        links.append({
            "platform": config["platform"],
            "display_name": config["display_name"],
            "icon": config["icon"],
            "search_url": url,
            "label": f"Browse {config['display_name']} Jobs",
            "search_query": role,
            "location": location,
        })
    return links
